### PR TITLE
Dnssec mitigations from audit

### PR DESCRIPTION
**Pull request by Arachnid**

### CODE REVIEW

File: contracts/dnssec-oracle/BytesUtils.sol [hunk 1]
```diff
@@ -76,10 +85,10 @@ library BytesUtils {
             if (a != b) {
                 // Mask out irrelevant bytes and check again
                 uint256 mask;
-                if (shortest > 32) {
+                if (shortest - idx >= 32) {
                     mask = type(uint256).max;
                 } else {
-                    mask = ~(2**(8 * (32 - shortest + idx)) - 1);
+                    mask = ~(2 ** (8 * (idx + 32 - shortest)) - 1);
```
Comment (makoto) [inline contracts/dnssec-oracle/BytesUtils.sol:91]: Is there any difference by changing from `32 - shortest + idx` to `idx + 32 - shortest`?

File: test/dnssec-oracle/TestRRUtils.sol [hunk 2]
```diff
@@ -58,6 +58,8 @@ contract TestRRUtils {
     bytes memory b_a_c  = hex'01620161016300';
     bytes memory ab_c_d = hex'0261620163016400';
     bytes memory a_c_d  = hex'01610163016400';
+  	bytes memory verylong1_eth = hex'223031323334353637383930313233343536373839303132333435363738393031613031303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637380365746800';
+  	bytes memory verylong2_eth = hex'2130313233343536373839303132333435363738393031323334353637383930316131303132333435363738393031323334353637383930313233343536373839303132333435363738393031323334353637380365746800';
```
Comment (makoto) [inline test/dnssec-oracle/TestRRUtils.sol:62]: indent is off

File: contracts/dnssec-oracle/DNSSECImpl.sol [hunk 3]
```diff
@@ -310,8 +309,11 @@ contract DNSSECImpl is DNSSEC, Owned {
             return false;
         }
 
-        return
-            algorithms[dnskey.algorithm].verify(keyrdata, data.rrset, data.sig);
+        Algorithm algorithm = algorithms[dnskey.algorithm];
+        if(address(algorithm) == address(0)) {
+            return false;
+        }
```
Comment (makoto) [inline contracts/dnssec-oracle/DNSSECImpl.sol:315]: which test is testing this condition ?

File: contracts/dnssec-oracle/BytesUtils.sol (hunk 1 above)
Comment (Arachnid) [reply contracts/dnssec-oracle/BytesUtils.sol:91]: See the original bug report - there's overflow issues with the previous ordering.

File: contracts/dnssec-oracle/DNSSECImpl.sol (hunk 3 above)
Comment (Arachnid) [reply contracts/dnssec-oracle/DNSSECImpl.sol:315]: I've just added one.

**Comment by Arachnid** [comment]
@makoto PTAL

File: test/dnssec-oracle/TestDNSSEC.js [hunk 4]
```diff
@@ -465,6 +465,142 @@ contract('DNSSEC', function(accounts) {
...
+              ttl: 3600,
+              data: ["Test"],
+            },
+          ],
+        })
+      ]), 'InvalidSignerName'
+    )      
+  })
+
+  it.only('should reject signatures with unknown algorithms', async function() {
```
Comment (makoto) [inline test/dnssec-oracle/TestDNSSEC.js:536]: Please remove .only

**Pull request merged**